### ISSUE TITLE
Universal links aren't being handled

### ISSUE THREAD

**Issue by justrealmilk** [pwa-builder/pwabuilder-ios#40]
After reading through #34, looking at recent commits, and experimenting, I can't seem to get this to work. I've checked out some tutorial material but they generally focus on URL schemes and the roles they play in a native application - not on using scenes and directing a web view to a URL!

For debugging, I'm running `xcrun simctl openurl booted https://<domain name>` from Terminal which successfully and reliably brings my app to foreground in the simulator. I'm an absolute novice but I chucked a `print()` into `SceneDelegate.application()` see if it gets hit and it doesn't. Just to check my prints were working, I also added one to `sceneDidBecomeActive()`. Unfortunately, it continues to fail on real devices, deployed through the store or via Xcode. 
https://bray.tech/.well-known/apple-app-site-association

My code is as the latest commits of the project template, which doesn't appear to be deployed to the prod generator yet.

**Comment by justrealmilk** [comment]
SceneDelegate.swift
```swift
    func scene(_ scene: UIScene, continue userActivity: NSUserActivity) {
        print("hello!")
        
        // Ensure we're trying to launch a link.
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
            let incomingURL = userActivity.webpageURL else {
            return
        }

        // Handle it inside our web view.
        let handledRequest = URLRequest(url: incomingURL)
        MyPWA.webView.load(handledRequest)
    }
```

Does successfully call and open the URL... But only when resuming the app (not a cold start) and it reloads the page as opposed to navigating a SPA to a different pathname.

**Comment by justrealmilk** [comment]
I've decided to use this as a general issue thread!

I use Edge as my default browser and noticed links aren't handled when tapped from other apps, including Discord... So I decided to do some searching and I found:

> I ran into the same issue. The info.plist was missing the LSApplicationQueriesSchemes entry with a https value.
> 
> For some reason, when Safari is the default browser, application.canOpenURL(url) returns true regardless of a LSApplicationQueriesSchemes entry. But when Chrome is the default browser it returns false and prints the following error in the console when that entry is missing.

https://developer.apple.com/forums/thread/665025

Additionally, https://stackoverflow.com/questions/64572364/ios-14-universal-links-broken-on-default-browser-other-than-safari

I've added this to my Info.plist
```DIFF
+ <key>LSApplicationQueriesSchemes</key>
+ <array>
+   <string>https</string>
+ </array>
```
But seen no change... Will update as I learn.

If I'm being a bother, just let me know... I figure that if I can't help by confidently making a PR, I can at least give someone a head start with my search results!

**Comment by JudahGabriel** [comment]
Hey @justrealmilk 

We had a fix for this via PR, but it wasn't deployed.

It's deployed as of today - give it a shot?

**Comment by JudahGabriel** [comment]
Nope, just testing it out, I can get universal link half working: clicking a link to my domain opens my PWA, but then my PWA doesn't actually handle the link and navigate. Investigating...

**Comment by JudahGabriel** [comment]
> Changing the function signature to this successfully opens the URL... But only when resuming the app (not a cold start) and it reloads the page as opposed to navigating a SPA to a different pathname. I'm hoping there's a better solution to this than calling callAsyncJavaScript and using the History API 😅

OK, for navigating the SPA, that should be simple enough, just changing the `MyApp.webView.load(...)` call to `MyApp.webView.evaluateJavaScript("location.href = '\(incomingURL)'")`. That works for my SPA which uses the Vaadin router - I suspect it'll work fine for other SPAs as well.

As for the other issue, this only firing if the app is running in the background, I'm investigating now.

As for the other issue, the

**Comment by JudahGabriel** [comment]
I'm beginning to wonder if the issue is the apple app-site-association file format itself. I understanding iOS 13 has a new format that is not compatible with the old format.

**Comment by JudahGabriel** [comment]
I have an updated Apple App Site Association (AASA) file, and now it appears it's working. Guidance forthcoming shortly...

**Comment by JudahGabriel** [comment]
OK, feel like I'm getting closer here. I think we need both functions in SceneDelegate.swift:

```swift
func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
   // If the app isn't running and we launch via universal link, we need to handle it here.
}

func scene(_ scene: UIScene, continue userActivity: NSUserActivity) {
   // If the app is already running and a universal link is clicked, handle it here. (Ideally, navigate in a SPA-friendly way)
}
```

The first one is giving me some trouble though. The following code will crash:

```swift
    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
        for userActivity in connectionOptions.userActivities {
               if let url = userActivity.webpageURL {
                   let urlReq = URLRequest(url: url)
                   MyApp.webView.load(urlReq)
                   break
               }
           }
    }
```

I'm wondering if it's because the webView isn't initialized at that point. It's hard to say because to test this, we have to kill the app first, which stops the debugger. Investigating...

**Comment by JudahGabriel** [comment]
Yes, indeed that is the problem. If I change the code to the following, it works! 

```swift
    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
        for userActivity in connectionOptions.userActivities {
               if let url = userActivity.webpageURL {
                   let urlReq = URLRequest(url: url)
                   // Navigate in 5 seconds. By then (race condition!) the web view will be initialized.
                   DispatchQueue.main.asyncAfter(deadline: .now() + 5.0, execute: {
                       MessianicChords.webView.load(urlReq);
                   })
                   break
               }
           }
        
    }
```

This works. But obviously we have a race condition. I'll have working code posted soon.

**Comment by JudahGabriel** [comment]
Woohoo! Got it working.

Step 1: `SceneDelegate.swift` should have handler for universal links when app is already running:

```swift
    // Handle universal links into our app when the app is already running.
    // This allows your PWA to open links to your domain, rather than opening in a browser tab.
    // See https://developer.apple.com/documentation/xcode/supporting-universal-links-in-your-app
    // This function is called when the app is already running and the user clicks a universal link.
    func scene(_ scene: UIScene, continue userActivity: NSUserActivity) {
        // Ensure we're trying to launch a link.
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
            let incomingURL = userActivity.webpageURL else {
            return
        }

        // Handle it inside our web view in a SPA-friendly way.
        let handledRequest = URLRequest(url: incomingURL)
        MyApp.webView.evaluateJavaScript("location.href = '\(incomingURL)'")
    }
```

Step 2: `SceneDelegate.swift` should have a handler for universal links when the app _isn't_ running:

```swift
    static var universalLinkToLaunch: URL? = nil; // If our app is launched with a universal link, we'll store it in this variable

    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
        
        // See if our app is being launched via universal link.
        // If so, store that link so we can navigate to it once webView is initialized.
        for userActivity in connectionOptions.userActivities {
               if let url = userActivity.webpageURL {
                   SceneDelegate.universalLinkToLaunch = url;
                   break
               }
           }
    }
```

Step 3: `ViewController.swift` should look to see if we were launched with a universal link, and if so, navigate to that instead of the root URL:

```swift
    // The loadRootUrl function already exists. Update it to the following code:
    @objc func loadRootUrl() {
        // Was the app launched via a universal link? If so, navigate to that.
        // If not, navigate to the app root URL.
        let universalLinkOrRootUrl = SceneDelegate.universalLinkToLaunch ?? rootUrl;
        MyApp.webView.load(URLRequest(url: universalLinkOrRootUrl));
    }
```

I'm committing these changes to the iOS template and deploying. Feel free to reopen this issue if you're still hitting problems.

**Comment by JudahGabriel** [comment]
I've published this code to PWABuilder.

**Comment by justrealmilk** [comment]
Awesome! A New Years treat. Unfortunately, the eval javascript doesn't work for `react-router` which I _think_ is the most popular routing package for React at the minute... I was able to reproduce the behaviour on trusty desktop browser whereby the whole app reloads from shortcuts or universal links and with cold and warm app starts. I started looking at the native history API but react-router doesn't listen for push and replace events there.
